src/libraries/System.IO.Pipelines/src/System/IO/Pipelines/PipeReader.cs

```diff
@@ -182,9 +191,19 @@ private async Task CopyToAsyncCore<TStream>(TStream destination, Func<TStream, R
 
                     while (buffer.TryGet(ref position, out ReadOnlyMemory<byte> memory))
                     {
-                        await writeAsync(destination, memory, cancellationToken).ConfigureAwait(false);
+                        FlushResult flushResult = await writeAsync(destination, memory, cancellationToken).ConfigureAwait(false);
+
+                        if (flushResult.IsCanceled)
+                        {
+                            ThrowHelper.ThrowOperationCanceledException_FlushCanceled();
+                        }
 
                         consumed = position;
+
+                        if (flushResult.IsCompleted)
```

Comment:
Do we have a test for this case? If you `PipeReader.CopyToAsync(PipeWriter)` and that pipe writer is complete before writing all of the data, and you copy the PipeReader to another Pipewriter and it resumes where it left off and continues copying.

Reply:
I have added a test proposal. Comments are welcome.
I have checked that without the fix this test case hangs... as expected.
However, I have a doubt around the update of the `consumed` position when the flush result is "completed". Indeed, by design, flush comes after write so data is considered as consumed (if not failed), but the [`Pipe.WriteAsync`](https://github.com/dotnet/runtime/blob/01b7e73cd378145264a7cb7a09365b41ed42b240/src/libraries/System.IO.Pipelines/src/System/IO/Pipelines/Pipe.cs#L959) method can return a completed FlushResult before consuming the provided data...
Do I miss something here?

Reply:
That's a really good point. Lets file another issue. There's basically no way to know if the data has been consumed or not. Great find!

Reply:
You shouldn't have this check if you're doing other important stuff in this loop (which you are).

Reply:
This was written during triage when I didn't have the full context. I've now read through most of the related issues.

It seems the real problem is that when you have a PipeWriter/Reader pair with arbitrary Pause/ResumeThreshold, you cannot observe the read loop completion just using PipeWriter. This has always been the case though. Checking FlushResult.IsCompleted doesn't change that.

You used by be able to use [https://github.com/dotnet/runtime/blob/01b7e73cd378145264a7cb7a09365b41ed42b240/src/libraries/System.IO.Pipelines/ref/System.IO.Pipelines.cs#L79-L80](PipeWriter.OnReaderCompleted) and a TCS for this, but that's been obsoleted. If we want to keep this API, we'd want it to be a Task anyway.

Reply:
I think we leave this as it was the existing behavior. Assuming it was consumed.